dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/trackedentity/hibernate/HibernateTrackedEntityInstanceStoreTest.java

```diff
@@ -0,0 +1,233 @@
+
+    @Mock
+    private OrganisationUnitStore organisationUnitStore;
+
+    @Mock
+    private SystemSettingManager systemSettingManager;
+
+    @InjectMocks
+    private HibernateTrackedEntityInstanceStore store;
+
+    @Captor
+    ArgumentCaptor<String> sqlQueryCaptor;
+
+    private TrackedEntityInstanceQueryParams params;
+
+    @BeforeEach
+    void setUp()
+    {
+        params = new TrackedEntityInstanceQueryParams();
+        params.setOrganisationUnits( Set.of( new OrganisationUnit( "orgUnitUid" ) ) );
+        params.setProgram( new Program( "programUid" ) );
+        params.setIncludeDeleted( false );
+        params.setOrganisationUnitMode( OrganisationUnitSelectionMode.SELECTED );
+        params.setTrackedEntityTypes( List.of( new TrackedEntityType( "ARV commodity", "ARV commodity" ) ) );
+
+        SqlRowSet sqlRowSet = Mockito.mock( SqlRowSet.class );
+        Mockito.when( jdbcTemplate.queryForRowSet( Mockito.anyString() ) ).thenReturn( sqlRowSet );
+    }
+
+    @Test
+    void whenOrderingByInactiveThenQueryAndSubqueryContainInactiveField()
+    {
+        //given
+        params.setOrders( List.of( new OrderParam( "inactive", OrderParam.SortDirection.DESC ) ) );
+
+        //when
+        store.getTrackedEntityInstanceIds( params );
+
+        //then
+        Mockito.verify( jdbcTemplate, Mockito.atLeast( 1 ) ).queryForRowSet( sqlQueryCaptor.capture() );
+        String sqlQuery = sqlQueryCaptor.getValue();
+        assertAll( () -> assertEquals( 2, countMatches( sqlQuery, ", TEI.inactive" ) ),
+            () -> assertEquals( 2, countMatches( sqlQuery, "ORDER BY tei.inactive DESC" ) ) );
+    }
+
+    @Test
+    void whenOrderingByUpdatedAtThenQueryAndSubqueryContainUpdatedAtField()
+    {
+        //given
+        params.setOrders( List.of( new OrderParam( "updatedAt", OrderParam.SortDirection.ASC ) ) );
+
+        //when
+        store.getTrackedEntityInstanceIds( params );
+
+        //then
+        Mockito.verify( jdbcTemplate, Mockito.atLeast( 1 ) ).queryForRowSet( sqlQueryCaptor.capture() );
+        String sqlQuery = sqlQueryCaptor.getValue();
+        assertAll( () -> assertEquals( 2, countMatches( sqlQuery, ", TEI.lastupdated" ) ),
+            () -> assertEquals( 2, countMatches( sqlQuery, "ORDER BY tei.lastUpdated ASC" ) ) );
+    }
+
+    @Test
+    void whenOrderingByUpdatedAtClientThenQueryAndSubqueryContainUpdatedAtClient()
+    {
+        //given
+        params.setOrders( List.of( new OrderParam( "updatedAtClient", OrderParam.SortDirection.ASC ) ) );
+
+        //when
+        store.getTrackedEntityInstanceIds( params );
+
+        //then
+        Mockito.verify( jdbcTemplate, Mockito.atLeast( 1 ) ).queryForRowSet( sqlQueryCaptor.capture() );
+        String sqlQuery = sqlQueryCaptor.getValue();
+        assertAll( () -> assertEquals( 2, countMatches( sqlQuery, ", tei.lastUpdatedAtClient" ) ),
+            () -> assertEquals( 2, countMatches( sqlQuery, "ORDER BY tei.lastUpdatedAtClient ASC" ) ) );
```

Comment:
While its interesting to write a unit test and assert on the SQL string I feel like we don't really know whether ordering works.

The fact that `ORDER BY tei.lastUpdatedAtClient ASC` appears twice in the SQL does not tell us whether the result is ordered by `lastUpdatedAtClient`.

Reply:
I agree, with this test we only assert that the element we want to order by is present in the query, and is present the expected amount of times, but we don't even know if the query will be successful.
My original idea was to write an integration test for this issue, but the convenience test doesn't support postgres and the controllerIntegrationTest have the problem of the async call with the transaction manager.

There was also the option of writing and e2e, but I don't think this feature is worth an e2e test. To me e2e should be used for features that if broken, deploys to production freeze until the issue is fixed. This one has been broken for a while and no one reported it...

That's why I wrote this unit test, to at least make sure that the param is present in the query, and as soon as we find a solution for the controllerIntegrationTest, we can write an integration test where we assert that the TEIs in the result are sorted as expected.

Reply:
But you can write an integration test using postgres for this store. This allows you to test that the you get the result in the correct order. Like in https://github.com/dhis2/dhis2-core/blob/373a5bb1b4a3ad14d44eefc57701dfdb716915fe/dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/EventExporterTest.java#L682-L696